### ISSUE TITLE
[Feature Request] A search box to search the cipboard entires.

### ISSUE THREAD

**Issue by qxlab** [TermiT/Flycut#46]
Something like this: http://sourceforge.net/projects/ditto-cp/

**Comment by jackdent** [comment]
+1

**Comment by MarkJerde** [comment]
Hi jackdent / qxlab,
How does this idea sound?

* From the bezel, user presses Command-F (the standard find keystroke) and begins typing.
* Some "Searching: <typed text>" message appears at the top or bottom of the bezel the same way as the "1 of 40" is displayed, perhaps in addition to that "1 of 40", perhaps including a count of number of matches.
* Searching begins immediately as each letter is typed, if possible.
* The first found match is displayed in the bezel.
* Pressing Command-G will move to the next found match, looping back to the start after the end is reached.
* Letters, numbers, symbols, and spaces typed will go into the search while arrows and other bezel navigation keystrokes and Command-F will close the search and continue with normal bezel use from the item that was displayed at the time of keystroke.
* The search will be case-insensitive.
* The search will only count each clipping once, so a clipping with the word "frog" in it 100 times won't tie up a search for "frog" through 100 Command-G presses.
* The bezel will not display any more of the clipping during search than it does outside of search, so for very large clippings that do not fit entirely in the bezel nothing will be done to "zoom" to the match but the user can rest assured that the text they were looking for is somewhere in what would be pasted.

This approach should be reasonably achievable.  Please let me know if you have any concerns or other ideas how it should work before I set to it.

Thanks,
Mark

**Comment by jackdent** [comment]
Hi Mark,

Thanks for responding! My one comment, although not so much a personal issue, would be that users who normally click in the top bar and select the item to paste with their cursor (as opposed to those who use the bezel) may miss out on the feature as it is currently described. Perhaps it might also be worth adding a search bar to the top bar widget? Evernote automatically positions the cursor in the field:

<img width="344" alt="screen shot 2015-08-14 at 3 54 50 pm" src="https://cloud.githubusercontent.com/assets/1302592/9285654/f0aa4dd8-429c-11e5-8a0d-9b847e444aac.png">

Thanks,
Jack

**Comment by MarkJerde** [comment]
Hi jackdent / qxlab,
Here's a beta build of Flycut with search capability.  I'm a bezel sort of guy, but I went an implemented it in the menu first and still need to add it to the bezel.  Go figure.  :-)

* The search is case-sensitive.
* Results are displayed in the menu up to the configured "Display in menu" limit.
* Results are updated immediately with each keystroke.
* Search box defaults to active when the menu is open, so the user can start typing without clicking the box.
* Backward compatibility is retained for any users who would click the menu and then use the arrow keys to navigate the menu (I'm not sure why they would, but someone probably has a surprising use case), so they will notice no change in what they are used to.
* This also includes my previous changes that are in a pending pull request and a fix for a five-second delay I was seeing while opening the Preferences (on a rather fast machine with SSD).  I'll hopefully have both of these into a pull request soon.

https://github.com/MarkJerde/LMBTFY/blob/master/Flycut-MenuSearchBeta.zip

Any feedback is appreciated.

Thanks!
Mark

**Comment by jackdent** [comment]
Hi Mark,

This is awesome! A couple of nits:

- The search bar does not allow you to select your entire search term using Cmd+A (nor can you Cmd+C to copy the search term).
- I'd put the search bar above the search terms: I think it's a more standard UI pattern.

Other than that, it looks great.

Thanks,
Jack

**Comment by MarkJerde** [comment]
Hi Jack,
I agree that having the search at the top of the menu is more standard, but I am concerned that putting it there would force some existing menu users to change their usage patterns for a feature that not all of them would be interested in.  I decided to balance the interests of those who would appreciate this feature with those who appreciate not having to change.  Since the search box is active when the menu is open, the difference between the two is minor to a user of the search; and since extra steps were taken to allow the arrow keys to navigate the menu identically to before the search was added it seems like a pretty good balance of everyone's interests to me.  I am very interested in hearing any more thoughts/ideas you have on this, though, since I'd like the final solution to be one that everyone is okay with.

Thanks for catching the Cmd+A / Cmd+C support.  I'll see if I can figure out how to do it.  For Cmd+C, do you think it would be better to add or not to add the search to the Flycut clipping history?  On the plus side, we all like having a clipping history so it seems natural to include it.  On the minus side, repeating the same search would then result in the copied search being included in the search results, which seems counterproductive.  I would venture to say that doing a bezel or menu paste into the search box is out of the question, since the bezel cannot open while the menu is open and selecting a menu item to paste into the menu would create an ambiguity with selecting a menu item to paste elsewhere.  Your thoughts?

I did notice that opening the menu and then closing it without selecting anything results in loss of focus for the application that had been in the foreground when the menu was opened.  I have a fix for this in the code I am getting ready to submit.

Thank you very much for helping iron out this feature,
Mark

**Comment by MarkJerde** [comment]
I added undo / redo / cut / copy / paste / select-all keystrokes to the search box.  Same link.
https://github.com/MarkJerde/LMBTFY/blob/master/Flycut-MenuSearchBeta.zip

After several failed attempts at intercepting keystrokes or adding hidden menu items it turned out the solution was plain easy.  Just add a default "edit menu" from Xcode to the menu, delete the unwanted items, and mark it as "hidden".  Doesn't display, but it still functions with no extra work.

I have this all ready to push to GitHub and submit a pull request, which I'll probably do in the next day or two unless any further problems / concerns / ideas are found.

Thanks!
Mark

**Comment by jackdent** [comment]
Hi Mark,

Apologies for the delayed response. I've played around with the latest build and have two more small bug/nit.

1. The menubar icon is no longer highlighted in blue when active, as it was before.
<img width="398" alt="screen shot 2015-09-04 at 12 54 07" src="https://cloud.githubusercontent.com/assets/1302592/9689451/1d2d3c36-5304-11e5-9d40-4a75e934c8e0.png">
<img width="179" alt="screen shot 2015-09-04 at 12 54 22" src="https://cloud.githubusercontent.com/assets/1302592/9689452/1d30e6c4-5304-11e5-9d68-b69133f55e00.png">

2. When the menubar dropdown is open and you press the shortcut to bring up the bezel, the bezel appears for ~ half a second but then quickly disappears. 

Finally, I agree that in general we want to minimise changes to how existing menu users interact with the app, but disagree that the search bar is one of these cases. I think that adhering to a standard UI pattern will help onboard new users, and that this concern trumps the desire to avoid introducing an extra key press for existing users. This is probably one of those vim/emacs or tabs/spaces questions, so I concede that these thoughts are purely those of personal opinion, rather than objectivity.

Best wishes,
Jack

**Comment by MarkJerde** [comment]
Hi Jack,
Fixed and fixed.  Also fixed a case where the menu would display non-filtered results while there was still text in the search box.  The build is in the same link as before.

I very much agree that standard UI patterns are beneficial.  It is why I use a Mac.  Given that a search box in a menu is a pretty big hack (outside of Apple-internal menus like Help and Spotlight, and thus not standard UI at all) and that most people won't have trouble recognizing the search box, I'd rather err on the side of keeping existing users happy.  I'm sure I just made a choice that will upset everyone with screen readers and other assistive devices.  I'll be glad to change it in a later patch if others weigh in.

Thanks again for all of your help vetting this.
Mark
